threat fields

dainperkins opened the pull request:
added threat fields for e.g. threat reports from IDS/IPS, NGFW, etc...

webmat commented on schemas/threat.yml:
```diff
@@ -0,0 +1,70 @@
+      type: keyword
+      short: Mitre ATT&CK Matrix tactic
+      description: >
+          The Mitre ATT&CK Matrix Tactic categorization of the identified threat 
```
@MikePaquette So do we want to go all in on Mitre ATT&CK and say that this field (and also the `.technique` field) are specifically the Mitre framework's tactic and technique?

Or should we make it optional? E.g.:

> The type of tactic used by this threat. You can use the Mitre ATT&CK Matrix Tactic categorization, for example.

My preference would be to make it optional.

MikePaquette replied:
@webmat @dainperkins I think ECS should not "require" the MITRE ATT&CK framework, but should allow it, and as @webmat suggests, making this explicit in the descriptions of the fields.

webmat commented:
Also, please make sure to run a full `make` before pushing.

Right now CI is failing because `schema.json` hasn't been updated. I think your merge of the master branch brought in a fix to the generation of this file, that's why your previous commit wasn't modifying that file.

MikePaquette commented on schemas/threat.yml:
```diff
@@ -0,0 +1,70 @@
+      short: Predicted accuracy of the threat identification
+      description: >
+        Typically a qualitative description of the confidence that the threat identification is correct
+      example: medium
```
@dainperkins I would prefer that we add fields `threat.tactic.reference` and `threat.technique.reference` to hold hyperlinks to reference material, rather than including the URL in the description field.  I also suggest that we use a `.name` and `.id` for each tactic and technique, as listed in https://github.com/elastic/ecs/pull/291, so that would give us:

- [ ] `threat.tactic.id`
- [ ] `threat.tactic.name`
- [ ] `threat.tactic.reference`
- [ ] `threat.technique.id`
- [ ] `threat.technique.name`
- [ ] `threat.technique.reference`

dainperkins commented:
Works for me Mike (was thinking the same thing but was unsure of which to
separate) - I'll update today

On Tue, Aug 13, 2019 at 9:46 AM Mike Paquette <notifications@github.com>
wrote:

> *@MikePaquette* commented on this pull request.
> ------------------------------
>
> In schemas/threat.yml
> <https://github.com/elastic/ecs/pull/505#discussion_r313404687>:
>
> > +
> +    - name: severity
> +      level: extended
> +      type: keyword
> +      short: Severity of the threat as identified by the observer/ threat intel feed
> +      description: >
> +        Typically a qualitative description of the risk associated with the observed threat or IOC
> +      example: high
> +
> +    - name: confidence
> +      level: extended
> +      type: keyword
> +      short: Predicted accuracy of the threat identification
> +      description: >
> +        Typically a qualitative description of the confidence that the threat identification is correct
> +      example: medium
>
> @dainperkins <https://github.com/dainperkins> I would prefer that we add
> fields threat.tactic.reference and threat.technique.reference to hold
> hyperlinks to reference material, rather than including the URL in the
> description field. I also suggest that we use a .name and .id for each
> tactic and technique, as listed in #291
> <https://github.com/elastic/ecs/pull/291>, so that would give us:
>
>    - threat.tactic.id
>    - threat.tactic.name
>    - threat.tactic.reference
>    - threat.technique.id
>    - threat.technique.name
>    - threat.technique.reference
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/ecs/pull/505?email_source=notifications&email_token=AJMZFSPDQFRPX3YLQ2D3LELQEK3LDA5CNFSM4IF32FJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBMWBKA#discussion_r313404687>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJMZFSLGFPBMHFWDWDPGEG3QEK3LDANCNFSM4IF32FJA>
> .
>

dainperkins commented:
Removed the lowercase, and updated the desc in the newest version


On Tue, Aug 13, 2019 at 9:46 AM Mike Paquette <notifications@github.com>
wrote:

> *@MikePaquette* commented on this pull request.
> ------------------------------
>
> In schemas/threat.yml
> <https://github.com/elastic/ecs/pull/505#discussion_r313404572>:
>
> > +      level: extended
> +      type: keyword
> +      short: Name or classification of the threat detected or reported
> +      description: >
> +        Threat name as reported by system or intel provider
> +
> +        The field value must be normalized to lowercase for querying. See
> +        the documentation section "Implementing ECS".
> +      example: ntp reserved mode denial of service
> +
> +    - name: tactic
> +      level: extended
> +      type: keyword
> +      short: Mitre ATT&CK Matrix tactic
> +      description: >
> +          The Mitre ATT&CK Matrix Tactic categorization of the identified threat
>
> @webmat <https://github.com/webmat> @dainperkins
> <https://github.com/dainperkins> I think ECS should not "require" the
> MITRE ATT&CK framework, but should allow it, and as @webmat
> <https://github.com/webmat> suggests, making this explicit in the
> descriptions of the fields.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/ecs/pull/505?email_source=notifications&email_token=AJMZFSJMJ4IVW6EFRTXFKBDQEK3KJA5CNFSM4IF32FJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBMWAEA#discussion_r313404572>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJMZFSIVBHCEQDT6K3IA6RTQEK3KJANCNFSM4IF32FJA>
> .
>

webmat commented on CHANGELOG.next.md:
```diff
@@ -11,6 +11,8 @@ Thanks, you're awesome :-) -->
 
 ### Added
 
+* added 'threat' fields for threat detections #505
```
Please capitalize "Added" :-)

dainperkins replied:
fixed

webmat commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+        name: text
+      short: Threat tactic 
+      description: >
+          Name of the type of tactic used by this threat. You can use the Mitre ATT&CK Matrix Tactic categorization, for example.
```
I think we need to step back a little more from ATT&CK in our descriptions, and clearly state that "tactic" is the broader category, and "technique" is a subcategory of how a tactic has been carried. 

This will be helpful for people using a different categorization framework.

dainperkins replied:
reorganized to move tactic &  technique top the bottom (grouping reported fields at the top vs potentially evaluated fields to the bottom)

webmat commented:
Suppose someone's looking at a Suricata alert like [this one](https://github.com/elastic/beats/blob/v7.3.0/x-pack/filebeat/module/suricata/eve/test/eve-alerts.log-expected.json#L42-L46):

```
        "suricata.eve.alert.category": "Attempted Information Leak",
        "suricata.eve.alert.gid": 1,
        "suricata.eve.alert.rev": 4,
        "suricata.eve.alert.signature": "ET POLICY curl User-Agent Outbound",
        "suricata.eve.alert.signature_id": 2013028,
```

How would you suggest the mapping is done?

I think `alert.signature_id` maps to `threat.id` and `alert.signature` maps to `threat.name`.

But should `alert.category` be mapped to `threat.technique.name`? Seems like we'd like to both allow people to keep their alert source's categorization, as it's good to have a 1:1 mapping with the source, when investigating what the alert means. But it would also be useful to let people map to a third party categorization framework. I'm sure there's already people who've mapped Suricata's alerts to the ATT&CK framework.

Or perhaps we should treat actual alerts as a separate entity than threats? Map the alert details to an eventual `alert` field set, and add the `threat` field set purely as a categorization scheme on top of the alert?

dainperkins commented:
I'd put id/name in as described.

I wouldn't be against adding threat.category to extend the basic threat
concept (as reported by the observer), then leaving the tactic and
technique as more specific places to record inferred tactic/technique
information (mapped to ATT&CK, or other internal threat analyzation)

At first I was thinking alert.category could map to threat.technique, but
then people have to decide [do I want to map observed information? Or do I
leverage an additional taxonomy?]  given the low impact I think having both
would be beneficial (if for no other reason that the ability to search for
vendor specific threat names, or external classification terms, instead of
one or the other)

I'll add a threat.category, and take a look at a couple other detection
types to see what else could go into the overall (risk is set, tho I have
yet to figure out how to add threat to the risk.yml fields as a possible
top level as threat doesn't exist in my risk branch...)

/d



On Thu, Sep 5, 2019 at 11:20 AM Mathieu Martin <notifications@github.com>
wrote:

> Suppose someone's looking at a Suricata alert like this one
> <https://github.com/elastic/beats/blob/v7.3.0/x-pack/filebeat/module/suricata/eve/test/eve-alerts.log-expected.json#L42-L46>
> :
>
>         "suricata.eve.alert.category": "Attempted Information Leak",
>         "suricata.eve.alert.gid": 1,
>         "suricata.eve.alert.rev": 4,
>         "suricata.eve.alert.signature": "ET POLICY curl User-Agent Outbound",
>         "suricata.eve.alert.signature_id": 2013028,
>
> How would you suggest the mapping is done?
>
> I think alert.signature_id maps to threat.id and alert.signature maps to
> threat.name.
>
> But should alert.category be mapped to threat.technique.name? Seems like
> we'd like to both allow people to keep their alert source's categorization,
> as it's good to have a 1:1 mapping with the source, when investigating what
> the alert means. But it would also be useful to let people map to a third
> party categorization framework. I'm sure there's already people who've
> mapped Suricata's alerts to the ATT&CK framework.
>
> Or perhaps we should treat actual alerts as a separate entity than
> threats? Map the alert details to an eventual alert field set, and add
> the threat field set purely as a categorization scheme on top of the
> alert?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/ecs/pull/505?email_source=notifications&email_token=AJMZFSMEPMQNZC6AEBSKVXDQIEPUPA5CNFSM4IF32FJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD57RBVA#issuecomment-528421076>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJMZFSI5IYCANJXTOAP46H3QIEPUPANCNFSM4IF32FJA>
> .
>

webmat commented:
@dainperkins another thing that occurred to me. It might be worth adding a field that states which kind of categorization is being done. Is it MITRE, are people shoving QRadar in tactic & technique, or another vendor specific thing.

So perhaps something like `threat.taxonomy`, or `threat.framework`, in which you could put values such as "mitre_attack", "qradar", etc...

I can guarantee that in a big enough organization, people will be more than one different taxonomy, because `[reasons]`. So being able to separate them from one another may be helpful.

dainperkins commented:
excellent idea

On Thu, Sep 5, 2019 at 1:58 PM Mathieu Martin <notifications@github.com>
wrote:

> @dainperkins <https://github.com/dainperkins> another thing that occurred
> to me. It might be worth adding a field that states which kind of
> categorization is being done. Is it MITRE, are people shoving QRadar in
> tactic & technique, or another vendor specific thing.
>
> So perhaps something like threat.taxonomy, or threat.framework, in which
> you could put values such as "mitre_attack", "qradar", etc...
>
> I can guarantee that in a big enough organization, people will be more
> than one different taxonomy, because [reasons]. So being able to separate
> them from one another may be helpful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/ecs/pull/505?email_source=notifications&email_token=AJMZFSLVKXFIYLQHG2DP3STQIFCDLA5CNFSM4IF32FJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6AE6DY#issuecomment-528502543>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJMZFSLJVNPVCO3LL26DECTQIFCDLANCNFSM4IF32FJA>
> .
>

dainperkins commented:
was also thinking sig revision would be a good add (suricata has it, as
does snort)


On Thu, Sep 5, 2019 at 1:58 PM Mathieu Martin <notifications@github.com>
wrote:

> @dainperkins <https://github.com/dainperkins> another thing that occurred
> to me. It might be worth adding a field that states which kind of
> categorization is being done. Is it MITRE, are people shoving QRadar in
> tactic & technique, or another vendor specific thing.
>
> So perhaps something like threat.taxonomy, or threat.framework, in which
> you could put values such as "mitre_attack", "qradar", etc...
>
> I can guarantee that in a big enough organization, people will be more
> than one different taxonomy, because [reasons]. So being able to separate
> them from one another may be helpful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/ecs/pull/505?email_source=notifications&email_token=AJMZFSLVKXFIYLQHG2DP3STQIFCDLA5CNFSM4IF32FJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6AE6DY#issuecomment-528502543>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJMZFSLJVNPVCO3LL26DECTQIFCDLANCNFSM4IF32FJA>
> .
>

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+      description: >
+        Typically a qualitative description of the observing systems confidence in the correct identification of a 
+        threeat (more prevalent in AI/ ML analysis) or confidence in a particular threat intel provider. The reported 
+        severity and confidence fields can be normalized to provide a threat.risk.score for the overall event.
```
will update pending risk fields going in

dainperkins replied:
fixed, will update once risk is in ECS

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+      short: Predicted accuracy of the threat identification provided by the observer/threat intel feed
+      description: >
+        Typically a qualitative description of the observing systems confidence in the correct identification of a 
+        threeat (more prevalent in AI/ ML analysis) or confidence in a particular threat intel provider. The reported 
```
fixed

dainperkins replied:
fixed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+      multi_field:
+        type: text
+        name: text
+      short: Mitre ATT&CK Matrix tactic Name
```
removed the references to Mitre from the short description

dainperkins replied:
fixed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+    - name: tactic.id
+      level: extended
+      type: keyword
+      short: Mitre ATT&CK Matrix tactic ID
```
removed mitre

dainperkins replied:
addressed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+    - name: tactic.reference
+      level: extended
+      type: keyword
+      short: Mitre ATT&CK Matrix tactic URL reference
```
removed mitre

dainperkins replied:
addressed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+      type: keyword
+      short: Predicted accuracy of the threat identification provided by the observer/threat intel feed
+      description: >
+        Typically a qualitative description of the observing systems confidence in the correct identification of a 
```
fixed

dainperkins replied:
fixed

dainperkins replied:
fixed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+      multi_field:
+        type: text
+        name: text
+      short: Mitre ATT&CK Matrix technique id
```
addressed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+    - name: technique.reference
+      level: extended
+      type: keyword
+      short: Mitre ATT&CK Matrix technique
```
addressed

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,131 @@
+      multi_field:
+        type: text
+        name: text
+      short: Mitre ATT&CK Matrix technique name
```
addressed

mbudge commented:
An "outcome" or "action" field could store actions take by the observer or agent.

For example
- Firewall blocking inbound or outbound connection
- Web proxy blocking outbound http request
- Anti-virus quarantining malware file
- Endpoint software blocking suspicious file
- IPS blocking connection

webmat commented:
I'd like to close on this soon, but there's something that's been bugging me about the proposal.

According to the field set's definition, it sounds like it would be fine to either use this to map one's alerts to a taxonomy, or to use the source alert categorization as is (e.g. the Suricata alert categories). I think this field set should purely focus on making it possible to map alerts to the user's chosen taxonomy. Can be mitre, can be another taxonomy.

Corollary to that, I think it would be important to work on a sister field set that is specifically meant to capture the generic concepts about alerts. Before we have both, I think having only "threat.*" will be confusing.

I think the absence of the "alert" field set is what led us to discussing the addition `.revision` in here, for example. I think tracking a revision makes sense if you're talking about the version of an alert, but not if you're mapping alerts to a taxonomy. I mean it would make sense to have it in `alert.revision`, not in `threat.revision`.

Here's how I would envision the two. Let's take the same example as above to illustrate:

```
        "suricata.eve.alert.category": "Attempted Information Leak",
        "suricata.eve.alert.gid": 1,
        "suricata.eve.alert.rev": 4,
        "suricata.eve.alert.signature": "ET POLICY curl User-Agent Outbound",
        "suricata.eve.alert.signature_id": 2013028,
```

Given the existence of an alert field set, the mapping I would see would be:

```
- message: "ET POLICY curl User-Agent Outbound" (suricata.eve.alert.signature)
- alert.id: 2013028 (suricata.eve.alert.signature_id)
- alert.category.id: 1 (suricata.eve.alert.gid)
- alert.category.name: "Attempted Information Leak" (suricata.eve.alert.category)
- alert.revision: 4 (suricata.eve.alert.rev)
```

Then if someone is using Mitre, they would have mapped the known Suricata alerts to the mitre taxonomy, and the resulting event would also have these populated:

```
- threat.framework: mitre_att&ck
- threat.tactic.id: TA0010
- threat.tactic.name: exfiltration
- threat.tactic.reference: https://attack.mitre.org/tactics/TA0010/
- threat.technique.id: T1048
- threat.technique.name: Exfiltration Over Alternative Protocol
- threat.technique.reference: https://attack.mitre.org/techniques/T1048
```

If the user is actually using a different taxonomy, all of `alert.*` would remain exactly the same, but only what's under threat would be changing to that other taxonomy.

webmat commented:
Heads up as well. Ray is working on a misp importer Filebeat module here https://github.com/elastic/beats/pull/13805. There's some commonalities.

Worth giving this a look, while pondering all this.

webmat commented on schemas/threat.yml:
```diff
@@ -0,0 +1,158 @@
+    - name: revision
+      format: string
+      level: extended
+      type: long
```
Just spotted this. It should be `keyword`. Not all revisions are purely numeric.

dainperkins commented:
Forgot to commit before pull req 1 today...

webmat commented on CHANGELOG.next.md:
```diff
@@ -11,6 +11,8 @@ Thanks, you're awesome :-) -->
 
 ### Added
 
+* Remnoved threat event fields #505
```
We simplified the PR, but this PR is still "Adding threat fields" :-)

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,91 @@
+- name: threat
+  title: Threat
+  group: 2
+  short: Fields describing detected or reported threats
```
done

dainperkins commented on schemas/threat.yml:
```diff
@@ -0,0 +1,91 @@
+    - name: technique.name
+      level: extended
+      type: keyword
+      multi_field:
```
done

MikePaquette commented on schemas/threat.yml:
```diff
@@ -0,0 +1,85 @@
+    - name: tactic.reference
+      level: extended
+      type: keyword
+      short: Therat tactic url reference.
```
"Therat" should be "Threat"

dainperkins replied:
fixed

webmat commented on schemas/threat.yml:
```diff
@@ -0,0 +1,85 @@
+   These fields are for users to classify alerts from all of their sources (e.g. IDS, NGFW, etc.) within a 
+   common taxonomy. The threat.technique.* are meant to capture the high level category of the threat 
+   (e.g. "exfiltration"). The threat.tactic.* fields are meant to capture which kind of approach is used by 
+   this detected threat, to accomplish the goal (e.g. "data compressed").
```
I asked you to adjust the fieldset definition, and I got tactic and technique reversed 😂 Sorry about that.

High level category is tactic, and technique is the actual method used. Please update it to undo my mistake:

> These fields are for users to classify alerts from all of their sources (e.g. IDS, NGFW, etc.) within a 
   common taxonomy. The threat.tactic.* fields are meant to capture the high level category of the threat 
   (e.g. "exfiltration"). The threat.technique.* fields are meant to capture the method used by 
   this detected threat, to accomplish the goal (e.g. "data compressed").

webmat commented:
Oops, that last LGTM was on the wrong PR 🤦‍♂
